pkg/gardenlet/operation/botanist/dualstackmigration.go

```diff
@@ -17,42 +17,74 @@ import (
+	}
+	return true
+}
+
 // DetermineUpdateFunction determines the update function for the shoot's status based on dual-stack migration readiness.
 func (b *Botanist) DetermineUpdateFunction(networkReadyForDualStackMigration bool, nodeList *corev1.NodeList) func(*gardencorev1beta1.Shoot) error {
+	shootIPFamilies := b.Shoot.GetInfo().Spec.Networking.IPFamilies
+	isSingleStack := len(shootIPFamilies) == 1
+	allNodesMigrated := nodesMigrated(nodeList, len(shootIPFamilies))
+
 	if networkReadyForDualStackMigration {
-		return func(shoot *gardencorev1beta1.Shoot) error {
-			shoot.Status.Constraints = v1beta1helper.RemoveConditions(shoot.Status.Constraints, gardencorev1beta1.ShootDualStackNodesMigrationReady)
-			constraint := v1beta1helper.GetCondition(shoot.Status.Constraints, gardencorev1beta1.ShootDNSServiceMigrationReady)
-			if constraint == nil {
-				constraint := v1beta1helper.GetOrInitConditionWithClock(b.Clock, shoot.Status.Constraints, gardencorev1beta1.ShootDNSServiceMigrationReady)
-				constraint = v1beta1helper.UpdatedConditionWithClock(b.Clock, constraint, gardencorev1beta1.ConditionProgressing, "DNSServiceMigration", "The shoot is migrating the kube-dns service.")
-				shoot.Status.Constraints = v1beta1helper.MergeConditions(shoot.Status.Constraints, constraint)
-			}
-			return nil
-		}
+		return b.createConstraintRemovalFunction(isSingleStack, allNodesMigrated)
 	}
+	return b.createNodeMigrationFunction(isSingleStack, allNodesMigrated)
+}
 
-	allNodesDualStack := true
-	conditionStatus := gardencorev1beta1.ConditionProgressing
-	conditionReason := "NodesNotMigrated"
-	conditionMessage := "The shoot is migrating to dual-stack networking."
-	for _, node := range nodeList.Items {
-		allNodesDualStack = allNodesDualStack && len(node.Spec.PodCIDRs) == 2
+func (b *Botanist) createConstraintRemovalFunction(isSingleStack, allNodesMigrated bool) func(*gardencorev1beta1.Shoot) error {
+	return func(shoot *gardencorev1beta1.Shoot) error {
+		if isSingleStack && !allNodesMigrated {
+			return nil // Don't remove constraint yet
+		}
+
+		shoot.Status.Constraints = v1beta1helper.RemoveConditions(shoot.Status.Constraints, gardencorev1beta1.ShootDualStackNodesMigrationReady)
+
+		if !isSingleStack && !b.hasDNSMigrationConstraint(shoot) {
+			b.addDNSMigrationConstraint(shoot)
+		}
+		return nil
 	}
-	if allNodesDualStack {
-		conditionStatus = gardencorev1beta1.ConditionTrue
-		conditionReason = "NodesMigrated"
-		conditionMessage = "All nodes were migrated to dual-stack networking."
+}
+
+func (b *Botanist) createNodeMigrationFunction(isSingleStack, allNodesMigrated bool) func(*gardencorev1beta1.Shoot) error {
+	status := gardencorev1beta1.ConditionProgressing
+	reason := "NodesNotMigrated"
+	message := "Migrating node pod CIDRs to match target network stack."
+
+	if allNodesMigrated || isSingleStack {
+		status = gardencorev1beta1.ConditionTrue
+		reason = "NodesMigrated"
+		message = "All node pod CIDRs migrated to target network stack."
 	}
```

Comment:
I think that the reason/message should be adapted in case of the dual-stack to single-stack migration because it does not fit in this scenario.

Maybe something like the following:

```suggestion
	if allNodesMigrated {
		status = gardencorev1beta1.ConditionTrue
		reason = "NodesMigrated"
		message = "All node pod CIDRs migrated to target network stack."
	} else if isSingleStack {
		status = gardencorev1beta1.ConditionTrue
		reason = "NodesMigrating"
		message = "Node pod CIDRs are currently migrated to target network stack."
	}
```